models/stock_location.py

```diff
@@ -54,12 +97,274 @@ def get_location(self, location_identifier):
+            pi_outcome.moves_inventory = \
+                self._process_pi_count_moves(request[PI_COUNT_MOVES])
+
+        if INVENTORY_ADJUSTMENTS in request:
+            pi_outcome.adjustment_inventory = \
+                self._process_inventory_adjustments(request[INVENTORY_ADJUSTMENTS])
+
+            for pre_adjs_req in request.get(PRECEDING_INVENTORY_ADJUSTMENTS, []):
+                self._process_single_preceding_adjustments_request(
+                    pre_adjs_req,
+                    pi_outcome.adjustment_inventory)
+
+        self._process_pi_datetime(pi_outcome)
+
+        return True
+
+    def _process_pi_datetime(self, pi_outcome):
+        current_time = datetime.now()
+        self.write({'u_date_last_checked': current_time})
+
+        if not pi_outcome.got_inventory_changes():
+            # No PI changes - the location is in a correct state
+            self.write({'u_date_last_checked_correct': current_time})
+
+    # PI Count Moves
+
+    def _process_pi_count_moves(self, count_moves, picking_type_id=None):
+        """
+            Returns the modified inventory in case consistent move
+            changes are specified in the request, None otherwise.
+
+            Creates an internal transfer by default, if no
+            picking type is specified.
+
+            Raises a ValidationError in case of invalid request.
+        """
+        Picking = self.env['stock.picking']
+
+        created_pickings = Picking.browse()
+
+        for count_move in count_moves:
+            created_pickings += \
+                self._create_pi_count_move_picking(count_move, picking_type_id)
+
+        return created_pickings if created_pickings.exists() else None
+
+    def _create_pi_count_move_picking(self, count_move, picking_type_id):
+        """
+            Validates the single PI count move request.
+            Creates and returns the related picking.
+
+            Raises a ValidationError in case of invalid request
+            or if the quants (either specified in the request or
+            related to a package) are already reserved.
+        """
+        Package = self.env['stock.quant.package']
+        Picking = self.env['stock.picking']
+        Quant = self.env['stock.quant']
+
+        # NB: locations are already validated
+        location_id = int(count_move['location_id'])
+        location_dest_id = int(count_move['location_dest_id'])
+        quants = None
+        quant_ids = []
+
+        if 'package_id' in count_move:
+            # NB: ignoring a possible 'quant_ids' entry, in case
+            # there's no previous schema validation
+            package = Package.get_package(int(count_move['package_id']))
+            quants = package._get_contained_quants()
+            quant_ids = quants.ids
+        elif 'quant_ids' in count_move:
+            quant_ids = [int(x) for x in count_move['quant_ids']]
+            quants = Quant.browse(quant_ids)
+
+            if len(quants) != len(quant_ids):
```

Comment:
len(quants) will always be equal to len(quant_ids), we need an exists() either here or at 242.

Reply:
Oh right, good catch. The recordset interface gives you different alternatives.
Perhaps we should write down best practices.